File: java/pqueue/src/main/java/com/js/datastructuresalgorithms/pqueue/JSPQueue.java
package com.js.datastructuresalgorithms.pqueue;

import java.util.ArrayList;
import java.util.List;
//import java.util.Map;
//import java.util.Set;

public class JSPQueue<T extends Comparable<T>> {

    //*******************
    // Private Members **
    //*******************
    private List<T> mList;
    //private Map<T, Set<Integer>> mMap;

    public int size() { return array().size(); }

    public void enqeue(T item) {
        array().add(item);
        if(array().size() != 1) {
            bubbleUp(array().size() - 1);
        }
    }

    public T dequeue() {
        T result = item(0);
        swap(0, array().size() - 1);
        array().remove(array().size() - 1);
        bubbleDown();
        return result;
    }

    public T peek() { return item(0); }

    //*******************
    // Private Members **
    //*******************

    private void bubbleDown() {
        int index = 0;
        int lesser = lesserChild(index);
        while(lesser != -1 && compare(index, lesser) > 0) {
            swap(index, lesser);
            index = lesser;
            lesser = lesserChild(index);
        }
    }

    private void bubbleUp(int index) {
        int parent;
        while(compare(parent = parent(index), index) > 0) {
            swap(parent, index);
            index = parent;
        }
    }

    private int lesserChild(int index)
    {
        int left = left(index);
        int right = right(index);
        int lesser = left < size() ? left : -1;
        if(lesser != -1 && right < size()) {
            lesser = compare(left, right) > 0 ? right : lesser;
        }
        return lesser;
    }

    private T item(int index) { return index >= array().size() ? null : array().get(index); }

    private void swap(int one, int two) {
        T tmp = array().get(one);
        array().set(one, array().get(two));
        array().set(two, tmp);
    }
    //                 0
    //         1                2
    //     3       4        5        6
    //   7   8   9   10  11   12  13    14

    private int compare(int one, int two) { return compare(one, item(two)); }

    private int compare(int one, T two) { return compare(item(one), two); }

    private int compare(T one, T two) { return one == null ? -1 : one.compareTo(two); }

    private int parent(int current) { return current == 1 ? 0 : (current - 1) / 2; }

    private int left(int current) { return current * 2 + 1; }

    private int right(int current) { return current * 2 + 2; }

    private List<T> array() { return mList == null ? mList = new ArrayList<>() : mList; }

    //private Map<T, Set<Integer>> map() { return mMap == null ? mMap = new HashMap<T, Set<Integer>>() : mMap; }

    //private Set set(T key)
    //{
    //    if (!map().containsKey(key)) {
    //        map().put(key, new TreeSet<>());
    //    }
    //    return map().get(key);
    //}
}

File: node/pqueue/src/types/JSPQueue.js
class JSPQueue {

    //*****************
    // Class Members **
    //*****************
    #mArray = [];

    size = () => this.#array().length;

    enqueue = (item) => {
        this.#array().push(item);
        if(this.size() > 1) {
            this.#bubbleUp(this.size() -1);
        }
    }

    dequeue = () => {
        const result = this.#array()[0];
        this.#swap(0, this.size() - 1);
        this.#array().pop();
        this.#bubbleDown();
        return result;
    }

    //             1
    //        3         4
    //     12   11

    peek = () => this.#array()[0];

    //debug = () => this.#array();

    //*******************
    // Private Members **
    //*******************

    #bubbleUp = (index) => {
        let parent;
        while(this.#compare(parent = this.#parent(index), index) > 0) {
            this.#swap(parent, index);
            index = parent;
        }
    }

    #bubbleDown = () => {
        let index = 0;
        let lesser = this.#lesserChild(index);
        while(lesser !== -1 && this.#compare(index, lesser) > 0) {
            this.#swap(index, lesser);
            index = lesser;
            lesser = this.#lesserChild(index);
        }
    }

    #lesserChild = (index) => {
        const left = this.#left(index);
        const right = this.#right(index);
        let lesser = left < this.size() ? left : -1;
        if(lesser !== -1 && right < this.size()) {
            lesser = Math.min(this.#item(left), this.#item(right)) === this.#item(left) ? left : right
        }
        return lesser;
    }

    #swap = (one, two) => {
        const tmp = this.#item(one);
        this.#array()[one] = this.#array()[two];
        this.#array()[two] = tmp;
    }

    #compare = (one, two) => this.#item(one) - this.#item(two);

    #item = index => this.#array()[index];

    #parent = (index) => index === 1 ? 0 : Math.floor((index - 1) / 2);

    #left = (index) => index * 2 + 1;

    #right = (index) => this.#left(index) + 1;

    #array = () => this.#mArray;

}

module.exports = JSPQueue;